python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,243 @@
+        self.NOT_CONTAINED = flagDefs.add(
+            "flag_not_contained", "Center not contained in footprint bounding box"
+        )
+        self.PARENT_SOURCE = flagDefs.add("flag_parent_source", "Parent source, ignored")
+        self.GALSIM = flagDefs.add("flag_galsim", "GalSim failure")
+        self.EDGE = flagDefs.add("flag_edge", "Variance undefined outside image edge")
+
+        # Flag handler.
+        self.flagHandler = FlagHandler.addFields(schema, name, flagDefs)
+
+        # Safe centroid extractor.
+        self.centroidExtractor = measBase.SafeCentroidExtractor(schema, name)
+
+    @classmethod
+    def getExecutionOrder(cls):
+        return cls.SHAPE_ORDER
+
+    def fail(self, record, error=None):
+        # docstring inherited.
+        self.flagHandler.handleFailure(record)
+        if error:
+            center = record.getCentroid()
+            self.log.debug(
+                "Failed to measure shape for %d at (%f, %f): %s",
+                record.getId(),
+                center.getX(),
+                center.getY(),
+                error,
+            )
+
+
+class HsmSourceMomentsConfig(HsmMomentsConfig):
+    # later we can put some doscstring
+    badMaskPlanes = pexConfig.ListField[str](
+        doc="Mask planes used to reject bad pixels.", default=["BAD", "SAT"]
+    )
+
+
+@measBase.register("ext_shapeHSM_HsmSourceMoments")
+class HsmSourceMomentsPlugin(HsmMomentsPlugin):
+    ConfigClass = HsmSourceMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None) -> None:
+        super().__init__(config, name, schema, logName)
+        # some flag thingy
+        # work in conjunction with fail()
+        self.centroidResultKey = afwTable.Point2DKey.addFields(
+            schema, name, "The centroid measured by HSM", "pixel"
+        )
+        self.shapeKey = afwTable.QuadrupoleKey.addFields(schema, name, doc="HSM source moments.")
+        if config.addFlux:  # search the correct attribute name in the codebase
+            self.fluxKey = measBase.FluxResultKey.addFields(schema, "flux", "The flux measured by HSM", "dn")
+
+    def measure(self, record, exposure):
+        """... in place"""
+        # source in c++ is record in py. exposure the same
+
+        # center = record.getCentroid()
+        center = self.centroidExtractor(record, self.flagHandler)
+
+        bbox = record.getFootprint().getBBox()
+        if bbox.getArea() == 0:
+            raise MeasurementError(self.NO_PIXELS.doc, self.NO_PIXELS.number)
+
+        if not bbox.contains(Point2I(center)):
+            raise MeasurementError(self.NOT_CONTAINED.doc, self.NOT_CONTAINED.number)
+
+        psfSigma = exposure.getPsf().computeShape(center).getTraceRadius()
+
+        # galsim
+        image_array = exposure[bbox].getImage().array
+        xmin, xmax = bbox.getMinX() + 1, bbox.getMaxX() + 1
+        ymin, ymax = bbox.getMinY() + 1, bbox.getMaxY() + 1
```

Comment:
It looks like that this `+1` is not required, and is perhaps causing the differences between the C++ implementation and pure-Python one.

Reply:
Done!